### PR TITLE
Feat: Improve the CLI output of the table diff command

### PR DESCRIPTION
**Pull request by izeigerman**
The following improvements have been introduced:
- Split `COMMON`, which represented the number of joined rows ,into `FULL MATCH` and `PARTIAL MATCH` to show the number of rows that are identical vs the number of rows with only  a partial match accordingly.
- Add percentage values for `FULL MATCHES` and `PARTIAL MATCHES`.
- Only display the number of rows if it's not 0.
- Add ability to set the number of decimal places to use when comparing floating point columns. This is set to 3 by default.

Screenshot 2024-05-21 at 10 46 42 AM

### CODE REVIEW

**Comment by sungchun12** [comment]
This is out of the scope for this PR, but is there a reason we don't show data diffs side by side like a git diff? We do it in the UI but not for the CLI. I imagine a lot of people want something similar in the CLI.

Example library that does this: https://github.com/paulfitz/daff

image

**Comment by izeigerman** [comment]
@sungchun12 , thanks, this is definitely on my mind! I even submitted an issue at the same time as this PR: https://github.com/TobikoData/sqlmesh/issues/2645. Any chance you can share this info there?

**Comment by treysp** [comment]
I naturally compare the row counts percentages and counts, and full match 6 is not 92.31% of 7. (I recognize why it's calculated the way it is, but I don't think we can expect users to.)

I think that calculation is going to be confusing unless we display percentages that sum to 100%.

**Comment by izeigerman** [comment]
@treysp I'm not sure we can, in good faith, show 100% when the number of rows doesn't match, right?

**Comment by treysp** [comment]
I meant show a percentage next to each count such that all the percentages sum to 100% - that way the user can tell how the denominator was calculated

**Comment by izeigerman** [comment]
@treysp ah, that's a good point, thanks. Updated my PR:

**Comment by sungchun12** [comment]
> @sungchun12 , thanks, this is definitely on my mind! I even submitted an issue at the same time as this PR: #2645. Any chance you can share this info there?

will do!

File: sqlmesh/core/table_diff.py [hunk 1]
```diff
@@ -187,9 +222,20 @@ def row_diff(self) -> RowDiff:
                     t_index.append(col)
             index_cols = list(dict.fromkeys(index_cols))
 
+            matched_columns = {
+                c: t for c, t in self.source_schema.items() if t == self.target_schema.get(c)
+            }
+
+            def _column_expr(name: str, table: str) -> exp.Expression:
+                if matched_columns[name].this in exp.DataType.FLOAT_TYPES:
+                    return exp.Round(
```
Comment (tobymao) [inline sqlmesh/core/table_diff.py:231]: exp.func("ROUND", exp.column(name, table), self.decimals)

File: sqlmesh/core/table_diff.py [hunk 2]
```diff
@@ -244,14 +289,31 @@ def name(e: exp.Expression) -> str:
                 .where(self.where)
             )
 
+            query = exp.select(
+                "*",
+                exp.Case()
+                .when(
+                    exp.and_(
+                        *[
+                            exp.column(f"{c}_matches").eq(exp.Literal.number(1))
```
Comment (tobymao) [inline sqlmesh/core/table_diff.py:298]: eq(1)

File: sqlmesh/core/table_diff.py [hunk 3]
```diff
@@ -244,14 +289,31 @@ def name(e: exp.Expression) -> str:
...
+                            for c in matched_columns
+                        ]
+                    ),
+                    exp.Literal.number(1),
```
Comment (tobymao) [inline sqlmesh/core/table_diff.py:302]: 1 and 0 for else_

File: sqlmesh/magics.py [hunk 4]
```diff
@@ -598,6 +598,12 @@ def create_external_models(self, context: Context, line: str) -> None:
         action="store_true",
         help="Show a sample of the rows that differ. With many columns, the output can be very wide.",
     )
+    @argument(
+        "--decimals",
+        type=int,
+        default=20,
```
Comment (tobymao) [inline sqlmesh/magics.py:604]: 3 not 20 right

**Review by tobymao — APPROVED** [review]

File: sqlmesh/magics.py (hunk 4 above)
Comment (izeigerman) [reply sqlmesh/magics.py:604]: ah, thanks for catching this. Not sure how this happened.

**Pull request merged**